### PR TITLE
New Maker API blocks (including `onBoardEvent`, `accelerometer`, `sensor`s)

### PR DESCRIPTION
**Pull request by bcjordan**

### CODE REVIEW

File: apps/src/applab/commands.js [hunk 1]
```diff
@@ -1839,6 +1839,12 @@ applabCommands.analogRead = function (opts) {
   return Applab.makerlabController.analogRead(opts.pin, opts.callback);
 };
 
+applabCommands.onBoardEvent = function (opts) {
+  //apiValidateType(opts, 'analogRead', 'pin', opts.pin, 'pinid');
```
Comment (bcjordan) [inline apps/src/applab/commands.js]: TODO: remove / validate event names?

File: apps/src/applab/commands.js [hunk 2]
```diff
@@ -1839,6 +1839,12 @@ applabCommands.analogRead = function (opts) {
   return Applab.makerlabController.analogRead(opts.pin, opts.callback);
 };
 
+applabCommands.onBoardEvent = function (opts) {
```
Comment (bcjordan) [inline apps/src/applab/commands.js:1842]: TODO: (maybe as separate commit): pull out into makerlab for easier integration with other app types

File: apps/src/dropletUtils.js [hunk 3]
```diff
@@ -614,7 +614,55 @@ exports.getAllAvailableDropletBlocks = function (dropletConfig, codeFunctions, p
     .concat(configuredBlocks);
 };
 
+/**
+ * @param {string} methodName name of method to get first param of
+ * @param {DropletBlock} block Droplet block, or undefined if in text mode
+ * @param {AceEditor} editor
```
Comment (bcjordan) [inline apps/src/dropletUtils.js:625]: TODO: document function throw, return value

File: apps/src/dropletUtils.js [hunk 4]
```diff
@@ -614,7 +614,55 @@ exports.getAllAvailableDropletBlocks = function (dropletConfig, codeFunctions, p
...
+    token = token.next;
+  } while (token);
+  return null;
+};
+
+function getFirstParamFromCode(methodName, code) {
+  var prefix = `${methodName}(`;
+  code = code.slice(code.lastIndexOf(prefix));
+  // quote, followed by param, followed by end quote, comma, and optional whitespace
+  const backslashEscapedRegex = `^${methodName}\\((['"])(.*)\\1,\\s*$`;
```
Comment (bcjordan) [inline apps/src/dropletUtils.js:659]: TODO: collapse into one regex

File: apps/src/makerlab/BoardController.js [hunk 5]
```diff
@@ -114,14 +116,26 @@ BoardController.prototype.reset = function () {
       if (component.off) {
         component.off();
       }
+      if (component.threshold) {
+        // Reset sensor thresholds to 1.0 in case changed.
+        component.threshold = 1;
```
Comment (bcjordan) [inline apps/src/makerlab/BoardController.js:133]: TODO: double check this is always the right starting value?

File: apps/src/makerlab/BoardController.js [hunk 6]
```diff
@@ -198,83 +216,6 @@ function deviceOnPortAppearsUsable(port) {
...
-    board: board,
-
-    /**
-     * Constants helpful for prototyping direct board usage
-     */
-    INPUT: 0,
-    OUTPUT: 1,
-    ANALOG: 2,
-    PWM: 3,
-    SERVO: 4
```
Comment (bcjordan) [inline apps/src/makerlab/BoardController.js:274]: TODO: pull up into BoardController instead

File: apps/src/makerlab/PlaygroundComponents.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
+/**
```
Comment (bcjordan) [inline apps/src/makerlab/PlaygroundComponents.js:1]: TODO: Resume review w/ Brad here

File: apps/src/makerlab/PlaygroundComponents.js [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+import {EventEmitter} from 'events'; // provided by webpack's node-libs-browser
+import LookbackLogger from './LookbackLogger';
+import _ from 'lodash';
+
+/**
+ * Initializes a set of Johnny-Five component instances for the currently
+ * connected Circuit Playground board.
+ *
+ * Components:
+ *   - will each have .start() called on Run
```
Comment (bcjordan) [inline apps/src/makerlab/PlaygroundComponents.js]: TODO: remove (not true anymore)

File: apps/src/makerlab/PlaygroundComponents.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+ *   - will each have .stop() called on Reset
+ *
+ * TODO(bcjordan): Import dependencies directly once conditional imports sorted.
+ *
+ * @returns {Object.<String, Object>} board components
+ */
+export function initializeCircuitPlaygroundComponents(io, board, five, PlaygroundIO) {
+  const colorLeds = Array.from({length: N_COLOR_LEDS}, (_, index) => new five.Led.RGB({
```
Comment (bcjordan) [inline apps/src/makerlab/PlaygroundComponents.js]: TODO: Change to use `_.count(N).map`

File: apps/src/makerlab/PlaygroundComponents.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  const soundSensor = new five.Sensor({
+    pin: "A4",
+    freq: 100
+  });
+  const buttonL = new five.Button('4');
+  const buttonR = new five.Button('19');
+  [buttonL, buttonR].forEach((button) => {
+    Object.defineProperty(button, 'isPressed', {
+      get: () => this.value === 1
```
Comment (bcjordan) [inline apps/src/makerlab/PlaygroundComponents.js]: TODO: fix (`button.value` or `function ()`)

File: apps/src/makerlab/PlaygroundComponents.js [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+  }
+}
+
+/**
+ * Adds `getAveragedValue` using LookbackLogger to a five.Sensor instance.
+ * @param {five.Board.fmap} fmap mapping function
+ * @param {five.Sensor} sensor
+ */
+const addSensorFeatures = (fmap, sensor) => {
+  /** @type {undefined} */
```
Comment (bcjordan) [inline apps/src/makerlab/PlaygroundComponents.js]: TODO: fix type

File: apps/src/makerlab/PlaygroundComponents.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    controller: "TINKERKIT",
+    pin: "A0",
+    freq: 100
+  });
+
+  [soundSensor, lightSensor, tempSensor].forEach((s) => {
+    addSensorFeatures(five.Board.fmap, s);
+  });
+
+  return _.merge(touchSensors, {
```
Comment (bcjordan) [inline apps/src/makerlab/PlaygroundComponents.js]: TODO: `_.assign({}, ...`

File: apps/src/makerlab/api.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+exports.analogRead = function (pin, callback) {
+  return Applab.executeCmd(null,
+      'analogRead',
+      {'pin': pin,
+        'callback': callback });
+};
+
+exports.onBoardEvent = function (component, event, callback) {
+  var extraArgs = Array.prototype.slice.call(arguments).slice(3);
```
Comment (bcjordan) [inline apps/src/makerlab/api.js]: TODO: remove.

File: apps/src/makerlab/dropletConfig.js [hunk 14]
```diff
@@ -19,50 +21,112 @@ const playSongConfig = {
...
+  accelerometer: ['change', 'data']
+};
+
+sensorVariables.forEach(s => eventDropdowns[s] = sensorEvents);
+touchSensorVariables.forEach(s => eventDropdowns[s] = touchEvents);
+
+/**
+ * Relies on `this` being the Droplet socket when in droplet mode, and, in
+ * text mode, this.parent being undefined.
+ * @param editor
```
Comment (bcjordan) [inline apps/src/makerlab/dropletConfig.js]: TODO: add type

File: apps/src/makerlab/dropletConfig.js [hunk 15]
```diff
@@ -19,50 +21,112 @@ const playSongConfig = {
...
+  {func: 'intensity', blockPrefix: colorLedBlockPrefix, category: CIRCUIT_CATEGORY, params: ['25'], tipPrefix: pixelType, modeOptionName: "*.intensity", objectDropdown: { options: colorPixelVariables }  },
+  {func: 'color', blockPrefix: colorLedBlockPrefix, category: CIRCUIT_CATEGORY, paletteParams: ['color'], params: ['"#FF00FF"'], paramButtons: { minArgs: 1, maxArgs: 3}, tipPrefix: pixelType, modeOptionName: "*.color", objectDropdown: { options: colorPixelVariables }  },
+
+  {func: 'buzzer', category: CIRCUIT_CATEGORY, type: 'readonlyproperty', noAutocomplete: true},
+  {func: 'buzzer.frequency', category: CIRCUIT_CATEGORY, paletteParams: ['frequency', 'duration'], params: ['500', '100'], paramButtons: { minArgs: 1, maxArgs: 2}},
+  {func: 'buzzer.note', category: CIRCUIT_CATEGORY, paletteParams: ['note', 'duration'], params: ['"A4"', '100'], paramButtons: { minArgs: 1, maxArgs: 2}},
+  {func: 'buzzer.off', category: CIRCUIT_CATEGORY},
+  {func: 'buzzer.stop', category: CIRCUIT_CATEGORY},
+  {func: 'buzzer.play', category: CIRCUIT_CATEGORY, paletteParams: ['song'], params: [JSON.stringify(playSongConfig)]},
```
Comment (bcjordan) [inline apps/src/makerlab/dropletConfig.js]: TODO: remove

File: apps/test/unit/dropletUtilsTest.js [hunk 16]
```diff
@@ -329,3 +329,36 @@ describe('filteredBlocksFromConfig', function () {
...
+    const code = "myProperty('element1', ";
+    assert.equal(getFirstParamFromCode('myProperty', code), 'element1');
+  });
+
+  it('works with double quotes', () => {
+    const code = 'myProperty("element1", ';
+    assert.equal(getFirstParamFromCode('myProperty', code), 'element1');
+  });
+
+  it('works with mix quotes', () => {
```
Comment (bcjordan) [inline apps/test/unit/dropletUtilsTest.js]: s/works/should not return anything with mix quotes

File: apps/test/unit/dropletUtilsTest.js [hunk 17]
```diff
@@ -329,3 +329,36 @@ describe('filteredBlocksFromConfig', function () {
...
+  it('works with no trailing space', () => {
+    const code = "myProperty('element1',";
+    assert.equal(getFirstParamFromCode('myProperty', code), 'element1');
+  });
+
+  it('works with multiple `myProperty`s', () => {
+    const code = "myProperty('element1', 'width', 100); myProperty('element2', ";
+    assert.equal(getFirstParamFromCode('myProperty', code), 'element2');
+  });
+  it('works with non-quoted strings (variable names)', () => {
```
Comment (bcjordan) [inline apps/test/unit/dropletUtilsTest.js:361]: - \n above

**Comment by islemaster** [comment]
LGTM :+1:

**Comment by bcjordan** [comment]
@islemaster updated further with changes from CR—PTAL!

File: apps/src/dropletUtils.js [hunk 18]
```diff
@@ -648,15 +655,11 @@ exports.getFirstParam = function (methodName, block, editor) {
 function getFirstParamFromCode(methodName, code) {
   var prefix = `${methodName}(`;
   code = code.slice(code.lastIndexOf(prefix));
-  // quote, followed by param, followed by end quote, comma, and optional whitespace
-  const backslashEscapedRegex = `^${methodName}\\((['"])(.*)\\1,\\s*$`;
-  const backslashNoQuoteRegex = `^${methodName}\\(([^"']*),\\s*$`;
+  // quote/empty, followed by param, followed by end quote/empty, comma, and optional whitespace
+  const backslashEscapedRegex = `^${methodName}\\((['"]?)(([^\\1]*)\\1,\\s*$`;
```
Comment (islemaster) [inline apps/src/dropletUtils.js]: :+1: Better!

File: apps/src/makerlab/PlaygroundConstants.js [hunk 19]
```diff
@@ -1,3 +1,5 @@
+import _ from 'lodash';
+
```
Comment (islemaster) [inline apps/src/makerlab/PlaygroundConstants.js:2]: Was just missing?

File: apps/src/makerlab/api.js [hunk 20]
```diff
@@ -1,44 +1,54 @@
 exports.pinMode = function (pin, mode) {
   return Applab.executeCmd(null,
       'pinMode',
-      {'pin': pin,
-        'mode': mode });
+      {
+        'pin': pin,
+        'mode': mode
+      });
```
Comment (islemaster) [inline apps/src/makerlab/api.js]: Using ES6 implicit object property names, could express as

``` javascript
return Applab.executeCmd(null, 'pinMode', {pin, mode});
```

**Comment by islemaster** [comment]
Nice cleanup!  LGTM.

File: apps/src/makerlab/PlaygroundConstants.js (hunk 19 above)
Comment (bcjordan) [inline apps/src/makerlab/PlaygroundConstants.js:2]: Yup

File: apps/src/makerlab/api.js (hunk 20 above)
Comment (bcjordan) [inline apps/src/makerlab/api.js]: Awesome :+1:

**Pull request merged**